nextcord/cog.py

```diff
@@ -0,0 +1,336 @@
+        TypeError
+            The function is not a coroutine function or a string was not passed as
+            the name.
+        """
+
+        if name is not MISSING and not isinstance(name, str):
+            raise TypeError(
+                f"Cog.listener expected str but received {name.__class__.__name__!r} instead."
+            )
+
+        def decorator(func: FuncT) -> FuncT:
+            actual = func
+
+            if isinstance(actual, staticmethod):
+                actual = actual.__func__
+
+            if not asyncio.iscoroutinefunction(actual):
+                raise TypeError("Listener function must be a coroutine function.")
+
+            actual.__cog_listener__ = True
+            to_assign = name or actual.__name__
+            try:
+                actual.__cog_listener_names__.append(to_assign)
+            except AttributeError:
+                actual.__cog_listener_names__ = [to_assign]
+
+            return func
+
+        return decorator
+
+    def process_app_cmds(self) -> None:
+        """Formats all added application commands with their callback."""
+        # TODO: Find better name, check conflicts with actual cogs.
+        for app_cmd in self.application_commands:
+            app_cmd.from_callback(app_cmd.callback)
+
+    @classmethod
+    def _get_overridden_method(cls, method: FuncT) -> Optional[FuncT]:
+        """Return None if the method is not overridden. Otherwise returns the overridden method."""
+        return getattr(method.__func__, "__cog_special_method__", method)
+
+    @_cog_special_method
+    def cog_application_command_check(self, interaction: Interaction) -> bool:
+        """A special method that registers as a :func:`.ext.application_checks.check`
+        for every application command and subcommand in this cog.
+
+        This function **can** be a coroutine and must take a sole parameter,
+        ``interaction``, to represent the :class:`.Interaction`.
+        """
+        return True
+
+    @_cog_special_method
+    async def cog_application_command_before_invoke(self, interaction: Interaction) -> None:
+        """A special method that acts as a cog local pre-invoke hook.
+
+        This is similar to :meth:`.ApplicationCommand.before_invoke`.
+
+        This **must** be a coroutine.
+
+        Parameters
+        ----------
+        interaction: :class:`.Interaction`
+            The invocation interaction.
+        """
+        pass
+
+    @_cog_special_method
+    async def cog_application_command_after_invoke(self, interaction: Interaction) -> None:
+        """A special method that acts as a cog local post-invoke hook.
+
+        This is similar to :meth:`.Command.after_invoke`.
+
+        This **must** be a coroutine.
+
+        Parameters
+        ----------
+        interaction: :class:`.Interaction`
+            The invocation interaction.
+        """
+        pass
+
+
+class ClientCog(Cog):
+    """An alias of :class:`Cog` for backwards compatibility reasons."""
```

Comment:
This might be better in nextcord.application_command because moving it here would break imports like `from nextcord.application_command import ClientCog` if people are currently doing that.

Feel free to ignore if you think it's not necessary.

Reply:
I could include the alias and add it to the `__all__` of `application_command.py`

Reply:
That should also work